### PR TITLE
[core][distributed] add zmq fallback for broadcasting large objects

### PR DESCRIPTION
**Pull request by youkaichao**
The input to vision language model contains images, which has variable length and can be quite large.

While the shared memory broadcast introduced in https://github.com/vllm-project/vllm/pull/5399 works fine for LLMs, later we find we often need to adjust the buffer size for vision language models.

Estimating the size upper bound can be difficult. To solve the problem, this PR adds a fallback option using zeromq. 
- When the object is small, we will use shared memory broadcast, which is fast and efficient.
- If the object is too large, we leave an overflow message in the shared memory, and send the data via zeromq, which can handle arbitrary sized data.

In addition, shared memory broadcast is limited to single node, while zeromq (socket-based) is not. Therefore, we can extend the broadcast to also work for cross-node settings. This PR extends the functionality.

cc @DarkLight1337 @ywang96 for vision language model related.

### CODE REVIEW

File: vllm/distributed/device_communicators/shm_broadcast.py [hunk 1]
```diff
@@ -236,29 +404,44 @@ def acquire_read(self):
...
+            if len(serialized_obj) >= self.buffer.max_chunk_bytes:
+                with self.acquire_write() as buf:
+                    buf[0] = 1  # overflow
+                self.local_socket.send(serialized_obj)
+            else:
+                with self.acquire_write() as buf:
+                    buf[0] = 0  # not overflow
+                    buf[1:len(serialized_obj) + 1] = serialized_obj
+        if self.n_remote_reader > 0:
+            self.remote_socket.send(serialized_obj)
```
Comment (youkaichao) [inline vllm/distributed/device_communicators/shm_broadcast.py:425]: this is the most critical part, previously, when object size is too large (e.g. large image data), vLLM will error. Now, we will fall back to zqm.

**Comment by DarkLight1337** [comment]
The idea sounds good, but I don't have much experience with cross-device communication, so I'll leave the review to someone who is more qualified.

File: vllm/distributed/parallel_state.py [hunk 2]
```diff
@@ -720,10 +721,12 @@ def init_world_group(ranks: List[int], local_rank: int,
...
 def init_model_parallel_group(
-        group_ranks: List[List[int]],
-        local_rank: int,
-        backend: str,
-        use_custom_allreduce: Optional[bool] = None) -> GroupCoordinator:
+    group_ranks: List[List[int]],
+    local_rank: int,
+    backend: str,
+    use_custom_allreduce: Optional[bool] = None,
+    use_mq_broadcaster: bool = False,
```
Comment (zhuohan123) [inline vllm/distributed/parallel_state.py:728]: Nit:
```suggestion
    use_message_queue_broadcaster: bool = False,
```
Also, how can a user turn this flag on?

**Review by zhuohan123 — APPROVED** [review]
I took a very brief look and the code in general LGTM. I didn't check the internal logic of the message queue-based broadcast, but the interface and code change looks good. Please let me know whether I need to look into any part more carefully.

File: vllm/distributed/parallel_state.py (hunk 2 above)
Comment (youkaichao) [reply vllm/distributed/parallel_state.py:728]: it is only used for tensor parallel now, where the broadcast happens.

Comment (youkaichao) [reply vllm/distributed/parallel_state.py]: fixed in [daa49cc](https://github.com/vllm-project/vllm/pull/6183/commits/daa49cca65ce5ba2d44ec2b17de14582a12f7dd4)

**Comment by WoosukKwon** [comment]
@youkaichao `pytest tests/distributed/test_shm_broadcast.py` worked on the AMD MI210 machine. Do you want other kinds of tests as well?

**Comment by youkaichao** [comment]
@WoosukKwon thanks for testing! I think it works because I added the dependency in `requirements-common.txt` , so all platforms can use it.

**Pull request merged**